tests/doc/test-levels.md

```diff
@@ -0,0 +1,150 @@
+exercises [HandleEquivocationEvidence](https://github.com/cosmos/cosmos-sdk/blob/2bec9d2021918650d3938c3ab242f84289daef80/x/evidence/keeper/infraction_test.go#L42) which contains many interactions with the staking
+keeper.
+
+#### Limitations
+
+Setting up a particular input state may be more challenging since the application is
+starting from "first principles".  Some of this may be addressed by good test fixture
+abstractions with testing of their own.  Tests may also be more brittle, and larger
+refactors could impact application initialization in unexpected ways with harder to
+understand errors.  This could also be seen as a benefit, and indeed the SDK's current
+integration tests were helpful in tracking down logic errors during earlier stages
+of app-wiring refactors.
+
+#### Proposal
+
+Maintain existing integration tests, but move them to `/tests/integration`.
+
+To help limit scope and complexity, prefer the smallest possible number of modules in
+application startup possible, i.e. don't depend on simapp.
+
+Maintain a *small* number of integration tests which use simapp directly to start up
+and test application operations.
+
+Integration tests outnumber end to end tests.
+
+### Simulation / generative testing 
+
+Simulations are a special case of integration tests where deterministically random module
+operations are executed against a running simapp.  No assertions are made for the state
+transitions resulting from module operations but any error will halt and fail the
+simulation.
+
+Modules must implement [AppModuleSimulation.WeightedOperations](https://github.com/cosmos/cosmos-sdk/blob/2bec9d2021918650d3938c3ab242f84289daef80/types/module/simulation.go#L31) to define their
+simulation operations.  Note that not all modules implement this which may indicate a
+gap in current simulation test coverage.
+
+Modules not returning simulation operations: 
+
+- `x/auth`
+- `x/capability`
+- `x/evidence`
+- `x/mint`
+- `x/params`
+
+A separate binary, [runsim](https://github.com/cosmos/tools/tree/master/cmd/runsim), is responsible for kicking off some of these tests and
+managing their life cycle.
+
+### Limitations
+
+- [A success](https://github.com/cosmos/cosmos-sdk/runs/7606931983?check_suite_focus=true) may take a long time to run, 7-10 minutes per simulation in CI. 
+- [Timeouts](https://github.com/cosmos/cosmos-sdk/runs/7606932295?check_suite_focus=true) sometimes occur on apparently successes without any indication why.
+- Useful error messages not provided on [failure](https://github.com/cosmos/cosmos-sdk/runs/7606932548?check_suite_focus=true) from CI, requiring a developer to run
+  the simulation locally to reproduce.
+
+#### Proposal
+
+No changes proposed at this time.
+
+## End to end tests
```

Comment:
An idea I have floated for a while is that we could have test suites that could be run either with integration (with mocked tendermint) or e2e fixtures (with real tendermint and many nodes). I'm not sure how easy this would be to set up or how much will there is around it, but it might be useful - integration fixtures could be used for quicker runs, e2e fixures could be used for more battle hardening

Reply:
I'm not understanding the proposal here. What would the benefit be of running the same test(s) with mocked vs real network?

Reply:
There might be some subtle things that are different. Maybe some non determinism issues could be caught?

Reply:
I like this conceptually, but I don't know if I can include it in the ADR proper right now since I'm just not sure how it would functionally work.  I'll add it to Future Discussions?